jsondiff.go

```diff
@@ -0,0 +1,886 @@
+					continue
+				}
+			}
+			// If the values are not equal, colorize them.
+			expectedOutput.WriteString(fmt.Sprintf("%s[%d]: %s\n", indent, i, red(serialize(aValue))))
+			actualOutput.WriteString(fmt.Sprintf("%s[%d]: %s\n", indent, i, green(serialize(bValue))))
+		}
+	}
+
+	// Return the resulting colorized differences for the expected and actual slices.
+	return expectedOutput.String(), actualOutput.String()
+}
+
+// serialize serializes a value to a pretty-printed JSON string.
+func serialize(value interface{}) string {
+	bytes, err := json.MarshalIndent(value, "", "  ")
+	if err != nil {
+		return "error"
+	}
+	return string(bytes)
+}
+
+// compare compares two values and writes the differences to the provided builders with optional colorization.
+// key: The key associated with the values being compared.
+// val1: The first value to compare.
+// val2: The second value to compare.
+// indent: The indentation string to use for formatting.
+// expect: The builder for the expected output.
+// actual: The builder for the actual output.
+// red, green: Functions to apply red and green colors respectively for differences.
+func compare(key string, val1, val2 interface{}, indent string, expect, actual *strings.Builder, red, green func(a ...interface{}) string) {
+	switch v1 := val1.(type) {
+	// Case for map[string]interface{} type
+	case map[string]interface{}:
+		// Check if the second value is also a map[string]interface{}
+		if v2, ok := val2.(map[string]interface{}); ok {
+			// Recursively compare and colorize maps
+			expectedText, actualText := compareAndColorizeMaps(v1, v2, indent+"  ", red, green)
+			expect.WriteString(fmt.Sprintf("%s\"%s\": %s\n", indent, key, expectedText))
+			actual.WriteString(fmt.Sprintf("%s\"%s\": %s\n", indent, key, actualText))
+			return
+		}
+		// If types do not match, write the key-value pairs with colors
+		writeKeyValuePair(expect, key, val1, indent, red)
+		writeKeyValuePair(actual, key, val2, indent, green)
+
+	// Case for []interface{} type
+	case []interface{}:
+		// Check if the second value is also a []interface{}
+		if v2, ok := val2.([]interface{}); ok {
+			// Recursively compare and colorize slices
+			expectedText, actualText := compareAndColorizeSlices(v1, v2, indent+"  ", red, green)
+			expect.WriteString(fmt.Sprintf("%s\"%s\": [\n%s\n%s]\n", indent, key, expectedText, indent))
+			actual.WriteString(fmt.Sprintf("%s\"%s\": [\n%s\n%s]\n", indent, key, actualText, indent))
+			return
+		}
+		// If types do not match, write the key-value pairs with colors
+		writeKeyValuePair(expect, key, val1, indent, red)
+		writeKeyValuePair(actual, key, val2, indent, green)
+
+	// Default case for other types
+	default:
+		// Check if the values are not deeply equal
+		if !reflect.DeepEqual(val1, val2) {
+			// Marshal values to pretty-printed JSON strings
+			val1Str, err := json.MarshalIndent(val1, "", "  ")
+			if err != nil {
```

Comment:
why are you not printing the error?

Reply:
considering this to be a sdk I thought of not adding print statements.